clients/go/compat.go

```diff
@@ -175,25 +208,34 @@ func (b WrappedBackend) CallContract(ctx context.Context, call ethereum.CallMsg,
 		}
 	} else {
 		leashBlockNumber := big.NewInt(0)
-		if blockNumber != nil {
-			leashBlockNumber.Sub(blockNumber, big.NewInt(1))
-		} else {
-			latestHeader, err := b.backend.HeaderByNumber(ctx, nil)
-			if err != nil {
-				return nil, fmt.Errorf("failed to fetch latest block number: %w", err)
-			}
-			leashBlockNumber.Sub(latestHeader.Number, big.NewInt(1))
-		}
-		header, err := b.backend.HeaderByNumber(ctx, leashBlockNumber)
+		header, err := b.backend.HeaderByNumber(ctx, blockNumber) // NB: blockNumber==nil will fetch the latest block.
 		if err != nil {
 			return nil, fmt.Errorf("failed to fetch leash block header: %w", err)
 		}
-		blockHash := header.Hash()
-		leash := NewLeash(header.Nonce.Uint64(), header.Number.Uint64(), blockHash[:], DefaultBlockRange)
+		// We will build a leash on the pre-last block.
+		blockHash := header.ParentHash
+		leashBlockNumber.Sub(header.Number, big.NewInt(1))
+		nonce, err := b.backend.PendingNonceAt(ctx, call.From)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch account nonce: %w", err)
+		}
+		fmt.Printf("BLOCK HASH: %x\n", blockHash)
+		headerJson, _ := json.Marshal(header)
+		fmt.Printf("BLOCK HEADER: %s\n", headerJson)
+		leash := evm.Leash{
+			Nonce:       nonce,
+			BlockNumber: leashBlockNumber.Uint64(),
+			BlockHash:   blockHash[:],
+			BlockRange:  DefaultBlockRange,
+		}
+		call.Gas = math.MaxInt64        // Must be non-zero for signing to work.
+		call.GasPrice = big.NewInt(0x1) // Must be non-zero for signing to work.
```

Comment:
Why are these overridden to some magic values? I would only override in case these are not set in `call` and these should be constants.